examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp

```diff
@@ -69,93 +75,232 @@ CHIP_ERROR LogProvider::GetLogForIntent(IntentEnum intent, MutableByteSpan & out
-    VerifyOrReturnError(dataSize, CHIP_ERROR_NOT_FOUND);
+    // In case of crash logs we can only mmap at max once, so check before doing anything
+    if (intent == IntentEnum::kCrashLogs)
+    {
+        VerifyOrReturnError(sCrashLogContext.isMapped == false, CHIP_ERROR_INCORRECT_STATE,
+                            ChipLogError(DeviceLayer, "Crash partition already mapped"));
+    }
 
-    ByteSpan * span = reinterpret_cast<ByteSpan *>(Platform::MemoryCalloc(1, sizeof(ByteSpan)));
-    VerifyOrReturnValue(span, CHIP_ERROR_NO_MEMORY);
+    LogContext * context = reinterpret_cast<LogContext *>(Platform::MemoryCalloc(1, sizeof(LogContext)));
+    VerifyOrReturnValue(context != nullptr, CHIP_ERROR_NO_MEMORY);
 
-    *span = ByteSpan(dataStart, dataSize);
+    CHIP_ERROR err = PrepareLogContextForIntent(context, intent);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err, Platform::MemoryFree(context));
 
     mLogSessionHandle++;
     // If the session handle rolls over to UINT16_MAX which is invalid, reset to 0.
     VerifyOrDo(mLogSessionHandle != kInvalidLogSessionHandle, mLogSessionHandle = 0);
 
-    outHandle                          = mLogSessionHandle;
-    mSessionSpanMap[mLogSessionHandle] = span;
+    outHandle                             = mLogSessionHandle;
+    mSessionContextMap[mLogSessionHandle] = context;
+
     return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR LogProvider::EndLogCollection(LogSessionHandle sessionHandle)
 {
     VerifyOrReturnValue(sessionHandle != kInvalidLogSessionHandle, CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrReturnValue(mSessionSpanMap.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnValue(mSessionContextMap.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
 
-    ByteSpan * span = mSessionSpanMap[sessionHandle];
-    mSessionSpanMap.erase(sessionHandle);
+    LogContext * context = mSessionContextMap[sessionHandle];
+
+    CleanupLogContextForIntent(context);
+    mSessionContextMap.erase(sessionHandle);
+    Platform::MemoryFree(context);
 
-    Platform::MemoryFree(span);
     return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR LogProvider::CollectLog(LogSessionHandle sessionHandle, MutableByteSpan & outBuffer, bool & outIsEndOfLog)
 {
     VerifyOrReturnValue(sessionHandle != kInvalidLogSessionHandle, CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrReturnValue(mSessionSpanMap.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
-
-    ByteSpan * span = mSessionSpanMap[sessionHandle];
-    auto dataSize   = span->size();
-    auto count      = std::min(dataSize, outBuffer.size());
-
-    VerifyOrReturnError(CanCastTo<off_t>(count), CHIP_ERROR_INVALID_ARGUMENT, outBuffer.reduce_size(0));
-
-    ReturnErrorOnFailure(CopySpanToMutableSpan(ByteSpan(span->data(), count), outBuffer));
+    VerifyOrReturnValue(mSessionContextMap.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
 
-    outIsEndOfLog = dataSize == count;
+    LogContext * context = mSessionContextMap[sessionHandle];
```

Comment:
please check null?

Reply:
Thanks for bringing it up, added the null checks at necessary places.